qiskit/ignis/verification/tomography/fitters/gateset_fitter.py

```diff
@@ -0,0 +1,662 @@
+            1) Use linear inversion to obtain an initial approximation.
+            2) Use gauge optimization to ensure the linear inversion results
+            are close enough to the expected optimization outcome to serve
+            as a suitable starting point
+            3) Use MLE optimization to obtain the final outcome
+        """
+        linear_inversion_results = self.linear_inversion()
+        n = len(self.gateset_basis.spam_labels)
+        E = self.default_measurement_op(n)
+        rho = self.default_init_state(n)
+        Gs = [self.gateset_basis.gate_matrices[label]
+              for label in self.gateset_basis.gate_labels]
+        Fs = [self.gateset_basis.spam_matrix(label)
+              for label in self.gateset_basis.spam_labels]
+        Gs_E = [linear_inversion_results[label]
+                for label in self.gateset_basis.gate_labels]
+        gauge_opt = GaugeOptimize(Gs, Gs_E, Fs, rho)
+        Gs_E = gauge_opt.optimize()
+        optimizer = GST_Optimize(self.gateset_basis.gate_labels,
+                                 self.gateset_basis.spam_spec,
+                                 self.probs)
+        optimizer.set_initial_value(E, rho, Gs_E)
+        optimization_results = optimizer.optimize()
+        return optimization_results
+
+
+class GaugeOptimize():
+    def __init__(self,
+                 Gs: List[np.array],
+                 initial_Gs_E: List[np.array],
+                 Fs: List[np.array],
+                 rho: np.array
+                 ):
+        """Initialize gauge optimizer fitter with the ideal and expected
+            outcomes.
+        Args:
+            Gs: The ideal expected gate matrices
+            initial_Gs_E: The experimentally-obtained gate approximations.
+            Fs: The SPAM circuit matrices
+            rho: The system's initial value (usually the |0> qubits)
+
+        Additional information:
+            Gauge optimization aims to find a basis in which the tomography
+            results are as close as possible to the ideal (noiseless) results
+
+            Given a gateset specification (E, rho, G1,...,Gn) and any
+            invertible matrix B, the gateset specification
+            (E*B^-1, B*rho, B*G1*B^-1,...,B*Gn*B^-1)
+            is indistinguishable from it by the tomography results.
+
+            B is called the gauge matrix and the goal of gauge optimization
+            is finding the B for which the resulting gateset description
+            is optimal in some sense; we choose to minimize the norm
+            difference between the gates found by experiment
+            and the "expected" gates in the ideal (noiseless) case.
+        """
+        self.Gs = Gs
+        self.Fs = Fs
+        self.initial_Gs_E = initial_Gs_E
+        self.d = np.shape(Gs[0])[0]
+        self.n = len(Gs)
+        self.rho = rho
+
+    def x_to_Gs_E(self, x: np.array) -> List[np.array]:
+        """Converts the gauge to the gateset defined by it
+                Args:
```

Comment:
This docstring has incorrect indent: 4 spaces should be removed from everything after the first line.

Reply:
Edit: Actually 8 spaces for this particular one

Reply:
Fixed for all